Blockers

mattigruener opened the pull request:
Hey John,

In the spirit of open source software, I thought we'd better discuss the rest of this publicly. I think it makes sense for us to chat about some of the outstanding things in person, though :) I'll list some remaining issues below as an agenda for our next video call.

- The biggest remaining issue: We designed the renderer backend with IPR in mind and what I have now doesn't work with the non-interactive code path. I believe for non-IPR renderings we're not holding on to the `ArnoldLights` and `ArnoldBlockers` long enough and they deregister themselves right after they've been created. I have a test that should fail and there's a comment outlining the issue. I think one way to solve this would be to hold a `std::unordered_set<ArnoldObjectPtr>` on `BlockersConnections` into which we insert `ArnoldLights` and `ArnoldBlockers` during registration and which we blow away when updating Arnold. That way lifetime would be guaranteed, but it's conceptually a little messy and maybe you have a better idea.

- We should discuss how much more visualisation is needed. Currently I visualise the `falloff` (except for spheres - see the comment in the code), but maybe we also want to visualise `density` similarly to what we're doing with `exposure` on lights? Maybe it's not important for a first version.

- I've cleaned up the `Blocker->ArnoldBlocker` relationship and the latter is now only responsible for providing an internal shader. The details are a little messy because what I feel should be a compile time error will currently throw an exception at runtime. There's a comment with the details. I'm also not sure if you like the way I'm promoting shader parameters to the wrapping class (see `Blocker::Blocker()`).

- There's a few more todo tags in the code, some of which we might want to address. Some others might be ok to leave for an eventual version 2... I need to add a few more tests as well, I think.

Looking forward to talking to you about this :) Cheerio for now,

Matti.

mattigruener commented:
Hey John,

just pushed a handful of fixes that we talked about this morning.

- Blockers now have a `loadShader()` method (and a custom serialiser)
- `BlockerConnections` holds on to `ArnoldLight` and `ArnoldObject` if needed. Could you have a look at whether you like the change to the `ArnoldRendererBase` ctor signature? Since `ArnoldRendererBase` doesn't know anything about the rendererType I opted for passing in the `BlockerConnections`. I think conceptually that's in line with the nodeDeleter being passed in?
- Using the above I fixed `ArnoldRenderTest`
- 'axis' plug now uses the right widget
- `EvaluateLightLinks` now computes the list of blocked lights by using the cache.

I'm expecting a lot more feedback once you start looking at things a bit closer... Thanks for your help thus far! :)

Cheerio!

Matti.

mattigruener commented:
Updated a test.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -317,6 +321,48 @@ class Instance;
+	struct Filters
+	{
+		ArnoldLight *light;
+		std::set<ArnoldBlocker*> blockers;
```
`unordered_set`

mattigruener replied:
Sorry, I messed up this one.

The change to the member is here: d55c78e
Here's where the rest is changed: 60efade130e878cec576ef060570928cf96e01c2

Will squash everything in anyway once we're done.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2026,13 +2244,132 @@ class ArnoldLight : public ArnoldObject
+
+void BlockerConnections::deregisterBlocker( ArnoldBlocker *blocker )
+{
+	for( auto &entry : m_connections )
```
`concurrent_hash_map` doesn't allow concurrent iteration, and iterating over _all_ lights seems like more work than we want to do anyway. Can we instead just pass in the lights that we want to deregister ourselves from?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2026,13 +2244,132 @@ class ArnoldLight : public ArnoldObject
+	// only happen when not iterating the map on another thread.
+	for( const std::string &lightName : deregistered )
+	{
+		m_connections.erase( lightName );
```
We _can_ actually call `erase()` concurrently, it's iteration that we can't do concurrently.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1974,18 +2174,13 @@ class ArnoldLight : public ArnoldObject
 
-			AiNodeSetArray( m_lightShader->root(), g_filtersArnoldString, linkedFilterNodes );
+			m_lightFilterShaders = arnoldAttributes->lightFilterShaders();
+			m_connections->registerLight( m_name, this );
```
This is re-registering the light every single time a light shader parameter is tweaked. Can't we just register once in the constructor?

mattigruener replied:
Only if we add another way to dirty a light, I think? It's a little misleading, but what this is essentially doing is signaling that the internal shader has been updated and that we'll need to set the blockers again. The fact that we throw away the `ArnoldShader` for every change means that we also throw away the blocker connections. Without this call, `updateFilters() `won't get called.

Do you want me to add a separate method that dirties a light? Do you see a better way of doing this? I don't remember why I opted for this way of doing things - maybe just to keep the `BlockerConnections` API small and you _could_ think of this as registering the new light (shader). Maybe it's sufficient to just add a comment in the code that explains it?

johnhaddon commented on python/GafferArnoldTest/ArnoldBlockerTest.py:
```diff
@@ -0,0 +1,80 @@
+import GafferSceneTest
+import GafferArnold
+
+class ArnoldBlockerTest( GafferSceneTest.SceneTestCase ) :
```
Neither of these tests pass, and `GafferArnoldTest.__init__` hasn't been updated to import them. I suggest we fix that, and beef the tests up to cover serialisation, assertSceneValid, keepExistingValues etc.

mattigruener replied:
This commit adjusts the test to the latest version of the rest of the blocker code: d168594
This one adds a few more tests: 236ddea

Any other you'd like me to add?

johnhaddon commented on python/GafferArnoldTest/ArnoldBlockerTest.py:
```diff
@@ -0,0 +1,80 @@
+
+		network = l["out"].attributes( "/blocker" )["ai:blocker"]
+		self.assertEqual( len( network ), 2 )
+		print network, dir(network)
```
Remove debug printing.

johnhaddon commented on python/GafferSceneTest/InteractiveRenderTest.py:
```diff
@@ -1541,6 +1541,166 @@ def testRendererContextVariable( self ):
 		image = IECoreImage.ImageDisplayDriver.storedImage( "myLovelySphereRenderedIn" + renderer )
 		self.assertTrue( isinstance( image, IECoreImage.ImagePrimitive ) )
 
+	def testLightFilters( self ) :
```
We're getting this error from the 3Delight tests now :

```
ERROR: testLightFilters (GafferDelightTest.InteractiveDelightRenderTest.InteractiveDelightRenderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/disk1/john/dev/build/gaffer/python/GafferSceneTest/InteractiveRenderTest.py", line 1548, in testLightFilters
    script["light"], colorPlug = self._createSpotLight()
AttributeError: 'InteractiveDelightRenderTest' object has no attribute '_createSpotLight'
```

Can you add `@unittest.skip()` to InteractiveDelightRenderTest, since we don't support light filters for 3Delight yet?

mattigruener replied:
15f15a3

mattigruener commented:
> Thanks Matti! Sorry for the delay in getting to this - it's a big 'un!

It is! Let's hope it's gonna be equally as helpful as time-consuming ;) It might take me a few days to get back to it - I have a bunch of other things on my plate at the moment. Thanks for your review! I did a first pass of reading through your comments and I think they all make sense. I hope I get to address them quickly!

mattigruener commented on include/GafferScene/Blocker.h:
```diff
@@ -0,0 +1,101 @@
+class GAFFERSCENE_API Blocker : public ObjectSource
+{
+
+	// Deriving classes are expected to provide a mechanism that constructs the
```
aca0a8b

mattigruener commented on include/GafferScene/Blocker.h:
```diff
@@ -0,0 +1,101 @@
+
+		void loadShader( const std::string &shaderName, bool keepExistingValues=false );
+
+		GafferScene::Shader *shaderNode();
```
aca0a8b

mattigruener commented on include/GafferScene/Blocker.h:
```diff
@@ -0,0 +1,101 @@
+
+		IE_CORE_DECLARERUNTIMETYPEDEXTENSION( GafferScene::Blocker, BlockerTypeId, ObjectSource );
+
+		Blocker( const std::string &name=defaultName<Blocker>(), GafferScene::Shader *shader = nullptr );
```
aca0a8b

mattigruener commented on src/GafferScene/Blocker.cpp:
```diff
@@ -0,0 +1,199 @@
+
+	addChild( shader );
+	addChild( new Plug( "parameters" ) );
+	addChild( new ShaderPlug( "__shader" ) );
```
aca0a8b

mattigruener commented on src/GafferScene/Blocker.cpp:
```diff
@@ -0,0 +1,199 @@
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	if( !shader )
```
aca0a8b

mattigruener commented on src/GafferScene/Blocker.cpp:
```diff
@@ -0,0 +1,199 @@
+
+void Blocker::loadShader( const std::string &shaderName, bool keepExistingValues )
+{
+	// Update internal shader
```
No _good_ reason. I ran into some issues back when I wrote this and have now looked into what was going on instead of just "making it work somehow". 

Here are two commits fixing things that gave me trouble before. Especially the first one as the parameter setup behaviour was so different from what we're doing on the OSL side. 11bb293 and e2cf6d1.

Assuming you are ok with those two fixes to existing code, these make things look more like an OSLLight unless I'm missing something: b5ff15b and 0adf089

Do you see any problems with my more fundamental changes?

mattigruener commented on src/GafferScene/Blocker.cpp:
```diff
@@ -0,0 +1,199 @@
+{
+	ObjectSource::affects( input, outputs );
+
+	if( parametersPlug()->isAncestorOf( input ) )
```
ea6294e

mattigruener commented on src/GafferScene/Blocker.cpp:
```diff
@@ -0,0 +1,199 @@
+}
+
+IECore::ConstCompoundObjectPtr Blocker::computeAttributes( const SceneNode::ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent ) const
+
```
fcc4d60

mattigruener commented on include/GafferScene/Blocker.h:
```diff
@@ -0,0 +1,101 @@
+#include "GafferScene/Shader.h"
+#include "GafferScene/ShaderPlug.h"
+
+#include "IECoreScene/ShaderNetwork.h"
```
fcc4d60

mattigruener commented on src/GafferArnoldModule/GafferArnoldModule.cpp:
```diff
@@ -78,6 +79,10 @@ BOOST_PYTHON_MODULE( _GafferArnold )
 	;
 
+	GafferBindings::DependencyNodeClass<ArnoldBlocker>()
+		.def( "loadShader", (void (ArnoldBlocker::*)( const std::string & ) )&ArnoldBlocker::loadShader )
```
942bd5b and 3bd5112

mattigruener commented on python/GafferArnoldUI/ArnoldBlockerUI.py:
```diff
@@ -0,0 +1,142 @@
+		# Activators need to live on the parent of the affected plug.
+		"parameters" : [
+
+			"layout:activator:shaderActivator", lambda parent : parent["geometry_type"].getValue() in ["box"],
```
c08ddbb3f7f7f177d7d3f89474d519cf7e2efd72

mattigruener commented on python/GafferArnoldUI/ArnoldBlockerUI.py:
```diff
@@ -0,0 +1,142 @@
+
+		],
+
+		"parameters.*" : [
```
54a74db

mattigruener commented:
Hey John :)

I updated the code with a first round of addressed feedback. I didn't get through the entirety of your notes yet, but maybe doing this in stages is actually better so I can squash in the stuff you are ok with and keep the commit history somewhat manageable.

I've indicated the respective commit in the comments and hope all of my updates make sense. The one about loading shader parameters onto a connected "parameters" plug was the missing piece to make things look more like OSLLight and I had missed that initially. What I have now should be ok?

Hopefully I can address the rest of the feedback tomorrow. Thanks for that last round of notes!

Matti.

mattigruener commented on python/GafferSceneTest/EvaluateLightLinksTest.py:
```diff
@@ -257,3 +257,40 @@ def testHash( self ) :
+
+		self.assertEqual(
+			set( map( str, evalNode["out"].attributes( '/group/sphere' )["blockedLights"] ) ),
+			set( ['/group/group/light', '/group/group/light1'] ) )
```
Do you prefer set displays over `set()` even though they can't create an empty set as needed above? I'm happy to change it, but I personally find it inconsistent with the previous check in this method.

mattigruener commented on include/GafferSceneUI/BlockerVisualiser.h:
```diff
@@ -0,0 +1,78 @@
+/// depending on their shader name (accessed using `IECore::Shader::getName()`). A
+/// factory mechanism is provided to map from this name to a specialised
+/// BlockerVisualiser.
+class GAFFERSCENEUI_API BlockerVisualiser : public IECore::RefCounted
```
368cb2c and f92e13b.

johnhaddon replied:
Sorry, I didn't make my point very clearly. What I'm trying to get at is that blockers are just light filters, so I was hoping that we didn't even need a BlockerVisualiser class, and StandardBlockerVisualiser could just derive from LightFilterVisualiser. Is there any reason we can't do that?

mattigruener commented on python/GafferSceneTest/EvaluateLightLinksTest.py:
```diff
@@ -257,3 +257,40 @@ def testHash( self ) :
+		attributes["in"].setInput( blocker["out"] )
+
+		mainGroup = GafferScene.Group( "MainGroup" )
+		mainGroup["in"].addChild( GafferScene.ScenePlug( "in1" ) )
```
1c7cf94

mattigruener commented on python/GafferSceneTest/EvaluateLightLinksTest.py:
```diff
@@ -257,3 +257,40 @@ def testHash( self ) :
+		light2 = GafferSceneTest.TestLight()
+
+		lightGroup = GafferScene.Group( "LightGroup" )
+		lightGroup["in"].addChild( GafferScene.ScenePlug( "in1" ) )
```
1c7cf94

mattigruener commented on python/GafferSceneTest/EvaluateLightLinksTest.py:
```diff
@@ -257,3 +257,40 @@ def testHash( self ) :
 
+	def testBlockerAttribute( self ) :
+
+		# \todo: Alright to just use a sphere here?
```
1c7cf94

mattigruener commented on python/GafferSceneTest/EvaluateLightLinksTest.py:
```diff
@@ -257,3 +257,40 @@ def testHash( self ) :
+		evalNode["in"].setInput( mainGroup["out"] )
+
+		self.assertEqual(
+			set( map( str, evalNode["out"].attributes( '/group/sphere' )["blockedLights"] ) ),
```
1c7cf94

mattigruener commented on python/GafferSceneTest/EvaluateLightLinksTest.py:
```diff
@@ -257,3 +257,40 @@ def testHash( self ) :
+
+		self.assertEqual(
+			set( map( str, evalNode["out"].attributes( '/group/sphere' )["blockedLights"] ) ),
+			set( [] ) )
```
1c7cf94

mattigruener commented on src/GafferSceneUI/StandardBlockerVisualiser.cpp:
```diff
@@ -0,0 +1,431 @@
+
+	IECoreGL::CurvesPrimitivePtr cube = new IECoreGL::CurvesPrimitive( IECore::CubicBasisf::linear(), /* periodic = */ true, vertsPerCurveData );
+	cube->addPrimitiveVariable( "P", IECoreScene::PrimitiveVariable( IECoreScene::PrimitiveVariable::Vertex, pData ) );
+	cube->addPrimitiveVariable( "Cs", IECoreScene::PrimitiveVariable( IECoreScene::PrimitiveVariable::Constant, new Color3fData( Color3f( 1.0f, 0.51f, 0.06f ) ) ) );
```
d234c41

I think this is better? It's quite a bit closer to what the lights look like - just a little more reddish.

mattigruener commented on src/GafferAppleseed/IECoreAppleseedPreview/Renderer.cpp:
```diff
@@ -1973,6 +1973,14 @@ class ProceduralRenderer : public AppleseedRendererBase
 		}
 
+
+		ObjectInterfacePtr lightBlocker( const std::string &name, const IECore::Object *object, const AttributesInterface *attributes ) override
```
I have combined these, but I'll look through everything again once you had time to check the new commits and I get to squash things in.

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -336,17 +382,19 @@ class ArnoldRendererBase : public IECoreScenePreview::Renderer
 	protected :
 
-		ArnoldRendererBase( NodeDeleter nodeDeleter, AtNode *parentNode = nullptr );
+		ArnoldRendererBase( NodeDeleter nodeDeleter, BlockerConnections *blockerConnections, AtNode *parentNode = nullptr );
```
60efade130e878cec576ef060570928cf96e01c2

Maybe we can talk about this when we skype the next time because I wasn't quite sure about the reference part of it. I've made it a regular smart pointer for now.

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1135,6 +1195,16 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+			return m_blockerShader.get();
+		}
+
+		IECore::ConstStringVectorDataPtr blockedLights() const
```
60efade130e878cec576ef060570928cf96e01c2

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2026,13 +2244,132 @@ class ArnoldLight : public ArnoldObject
 		const AtNode *m_parentNode;
 		ArnoldShaderPtr m_lightShader;
+		BlockerConnections *m_connections;
+		std::vector<ArnoldShaderPtr> m_lightFilterShaders;
```
60efade

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -317,6 +321,48 @@ class Instance;
+
+/// The BlockerConnections can be used to record changes regarding connections
+/// between blockers and lights and to communicate them to Arnold.
+class BlockerConnections : public IECore::RefCounted
```
d55c78ef9fad143a7812c7699bccc7b4b85ec160

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2026,13 +2244,132 @@ class ArnoldLight : public ArnoldObject
+				a->second.dirty = false;
+			}
+		},
+		tbb::auto_partitioner() );
```
60efade130e878cec576ef060570928cf96e01c2

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2026,13 +2244,132 @@ class ArnoldLight : public ArnoldObject
+	}
+	else
+	{
+		throw IECore::Exception( boost::str( boost::format( "Can not deregister non-existing light \"%s\"" ) % lightName.c_str() ) );
```
60efade

mattigruener commented:
Hey John,

I believe I've addressed all feedback. It's quite a bit and I've tried to keep the commits somewhat readable, but if you want me to reshuffle things to make it easier for you to review just let me know please :)

There's a few things that might require a quick conversation. Let's find some time to talk this week? :)

Thanks!

Matti.

mattigruener commented:
I've added another two commits to the pile. Should address the things we talked about today and I'll merge them in once we're ready.

johnhaddon commented:
Thanks Matti - this is looking good. Could you take care of my latest BlockerVisualiser and `auto &` comments and squash things down to the minimal set of commits we'll want in the history?

Where are we at on the Blocker->LightFilter rename? I'm feeling fairly sold on it, but if you want to make a case against then maybe do that before we go changing everything?

mattigruener commented:
Thanks for the latest review! :)

> Where are we at on the Blocker->LightFilter rename? I'm feeling fairly sold on it, but if you want to make a case against then maybe do that before we go changing everything?

I certainly don't feel strongly about it. My only concern would be that in some spots it might actually be necessary to differentiate between the light filters in light space and those in world space. I'm happy to go ahead and rename things after I squashed things down to see how well that works out. I'll do that today and we can look at it early next week :)

mattigruener commented:
Hey John,

I did some late renaming and hope it all makes sense. I'll check again on Monday morning with a fresh brain, but thought I'd put things up already anyway.

I also had another stab at the visualiser and hope this is closer to what you had in mind? I still wasn't sure if this is exactly what you meant but it feels fairly clean? I had to do that one change in LightFilterVisualiser, but that's probably ok? All those changes are in 24f9434, 
1e63e77 and e9beec4. Note that I didn't do any renaming for the files that get deleted anyway. I just included them for now to have a record of what the latest round of visualiser changes actually did.

Everything else is squashed in. Once we consider the renaming successful, I'll squash those commits in as well and then rebase everything onto current master.

Matti.

johnhaddon commented:
Could you fix the conflicts on this one please Matti?

johnhaddon commented:
I've been testing this out interactively and it has all been working nicely - thanks! I've got a few initial comments/questions from a useability perspective :

- I think having a `filteredLights` plug directly on the LightFilter node would make it easier to use/discover the linking functionality. I think we discussed this idea before - it would just let you set the filtered lights attribute without having to use a separate StandardAttributes node. Making this easy seems doubly important given that we've chosen to have filters not be linked to anything by default.
- Having got my linking set up, I was confused that I still wasn't getting any blocking effect. Perhaps we should provide a user default of 1 for Arnold's `light_blocker.density` parameter?
- It seems that the geometry type and falloff settings for Arnold's `light_blocker` are ignored as soon as a shader is plugged in. I presume this is just native Arnold behaviour? Perhaps we can help people out by greying out all those options based on having the connection?
- It'd be great if we could add the `ui:scene:acceptsSetExpression` metadata to the StandardAttributes linkedLights and filteredLights plugs.

mattigruener commented:
Thanks John! All of these make a lot of sense. I hope I can put up code for all of these later today :)

mattigruener commented:
The previous push contains only fixes that resolve conflicts. Sorry, I hadn't noticed them before.

mattigruener commented:
Pushed a few commits that hopefully address your latest feedback.

johnhaddon commented on src/GafferScene/LightFilter.cpp:
```diff
@@ -137,11 +149,17 @@ IECore::ConstObjectPtr LightFilter::computeSource( const Context *context ) cons
 void LightFilter::hashAttributes( const SceneNode::ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
 {
 	h.append( shaderPlug()->attributesHash() );
+	filteredLightsPlug()->hash( h );
```
We're missing a call to the base class hash in this function. It's a bit complicated because ObjectSource has a specialised implementation that isn't what we want, so the right thing to do is call through to SceneNode. Could you add this at the top of this function?

```
// We must call the base class before appending to the hash, but our direct
// base class (ObjectSource) is set up with a hardcoded hash suitable only
// for outputting empty attributes. Call directly to our SceneNode ancestor instead.
SceneNode::hashAttributes( path, context, parent );
```

mattigruener replied:
Done here: 237d6e0

I'll assume the missing hash in the arguments was a typo and not some intentional thing I just don't understand :)

johnhaddon commented on python/GafferSceneUI/SetUI.py:
```diff
@@ -193,7 +193,7 @@ def __setsPopupMenu( menuDefinition, plugValueWidget ) :
 
 	node = plug.node()
-	if isinstance( node, GafferScene.Filter ) :
+	if isinstance( node, (GafferScene.Filter, GafferScene.ObjectSource) ) :
```
What's the intention here? The general meaning of the "ui:scene:acceptsSetExpression" metadata is that a menu should be shown for sets available as inputs to the node (or filter). In this case, there are no such sets, and all we're achieving is to list sets that are output from the ObjectFilter, none of which are useful for the filteredLights attribute as far as I can see. There's also an open ticket at IE that is going to require some changes to `__downstreamNodes()` so that it becomes much more filter-specific. Unless I'm missing something, we should drop this and also remove the "ui:scene:acceptsSetExpression" metadata from the LightFilter node.

mattigruener replied:
My intention was to provide the menu in a graph like this. We don't have to, but I thought it'd be a nice touch and easy to do. I wasn't aware of the other ticket.

```
import Gaffer
import GafferArnold
import GafferScene
import IECore
import imath

__children = {}

__children["light_blocker"] = GafferArnold.ArnoldLightFilter( "light_blocker" )
parent.addChild( __children["light_blocker"] )
__children["light_blocker"].loadShader( "light_blocker" )
__children["light_blocker"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"] = GafferScene.Parent( "Parent" )
parent.addChild( __children["Parent"] )
__children["Parent"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["spot_light"] = GafferArnold.ArnoldLight( "spot_light" )
parent.addChild( __children["spot_light"] )
__children["spot_light"].loadShader( "spot_light" )
__children["spot_light"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Set"] = GafferScene.Set( "Set" )
parent.addChild( __children["Set"] )
__children["Set"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["light_blocker"]["parameters"]["density"].setValue( 1.0 )
__children["light_blocker"]["__uiPosition"].setValue( imath.V2f( 0.94999975, -1.66406333 ) )
__children["Parent"]["in"].setInput( __children["light_blocker"]["out"] )
__children["Parent"]["parent"].setValue( '/' )
__children["Parent"]["child"].setInput( __children["Set"]["out"] )
__children["Parent"]["__uiPosition"].setValue( imath.V2f( 12.9491501, -9.82812595 ) )
__children["spot_light"]["__uiPosition"].setValue( imath.V2f( 22.0491486, 6.5 ) )
__children["Set"]["in"].setInput( __children["spot_light"]["out"] )
__children["Set"]["name"].setValue( 'bestLights' )
__children["Set"]["__uiPosition"].setValue( imath.V2f( 22.0491486, -1.6640625 ) )

del __children

```

johnhaddon replied:
Hmm, it is handy in that example, but it's really fragile - stick a Group under the light_blocker and before the Parent and it doesn't work at all. I do think we're better off backing out of this one I'm afraid...

mattigruener replied:
Done in 3125b5c. Let me know if you want me to drop that commit entirely.

johnhaddon commented on src/GafferScene/LightFilter.cpp:
```diff
@@ -137,11 +149,17 @@ IECore::ConstObjectPtr LightFilter::computeSource( const Context *context ) cons
+	IECore::CompoundObjectPtr result = new IECore::CompoundObject;
+
+	result->members() = shaderPlug()->attributes()->members();
+	result->members()[g_filteredLightsAttributeName] = new IECore::StringData( filteredLightsPlug()->getValue() ); 
```
Is this definitely what we want when the string is empty? Perhaps it's better to not make the attribute in that case, so that several default-valued blockers could be grouped, and their linking set once with a StandardAttributes at the group level?

mattigruener replied:
Good point - 237d6e0. I've also extended the test in 8803536.

johnhaddon commented on python/GafferSceneUI/LightFilterUI.py:
```diff
@@ -52,6 +52,18 @@
+			all lights that haven't been excluded from that set.
+			""",
+
+			"ui:scene:acceptsSetExpression", True,
```
As mentioned in an earlier comment, I'm not sure this is helpful. Perhaps a preset for "defaultLights" might be a useful alternative? And what do you think to putting this in a "Light Linking" tab, analogous to the one on lights? I think it might be nice to separate it out from the regular parameters.

mattigruener replied:
I wouldn't mind the separated tab. I've put it right in the main one because we'd said that we'd like the plug to be easily discoverable as we don't link to anything by default. I'm happy to squirrel it away in its own tab, though :) I guess we need to discuss how much we consider the two mechanisms as separate things as per my previous comment before deciding what to call the tab?

johnhaddon replied:
I almost think a separate tab draws more attention to it, rather than hiding it amongst the shader parameters. And yeah, it should be definitely considered alongside the StandardAttributes question.

mattigruener replied:
I've done it here: 5f4f1dd. Called it "Light Linking" for now as I have a feeling that's what we'll settle on. We can change it later if needed.

johnhaddon commented on python/GafferArnoldUI/ArnoldLightFilterUI.py:
```diff
@@ -101,6 +117,34 @@ def __activateRoundness( parent ) :
-	}
+		"parameters.width_edge" : [
+
+			"layout:visibilityActivator", "falloffActivator",
```
Using "visibilityActivator" is making a confusingly empty Falloff section. I suggest we just use a regular "activator" for now, although at some point we should perhaps hide empty sections.

mattigruener replied:
Done here: de31e60

johnhaddon commented on python/GafferArnoldUI/ArnoldLightFilterUI.py:
```diff
@@ -56,9 +56,17 @@ def __activateShader( parent ) :
 
+def __activateFalloff( parent ) :
+
+	if parent["shader"].getInput() :
```
This isn't the behaviour I'm seeing out of Arnold (5.2). If I plug in a shader, it seems that the geometry type is ignored entirely, as well as the falloff. Do you see something different?

mattigruener replied:
Yeah, what I'm seeing is that when I use a "box", the `width_edge` and `height_edge` are used to scale the filter and the ramp is actually doing a rampy falloff. I'm not saying the edges make a lot of sense, but if they were set to something other than 0 and then I'd just go and deactivate the plugs, people were stuck with their previous settings that are now interpreted as scale...

johnhaddon replied:
Maybe I haven't described what I'm seeing well enough. I'm not using a ramp, I'm plugging a noise into the shader input. As soon as I do that, all geometry/falloff controls appear to be ignored entirely.

```
import Gaffer
import GafferArnold
import GafferImage
import GafferOSL
import GafferScene
import IECore
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 54, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["SceneReader"] = GafferScene.SceneReader( "SceneReader" )
parent.addChild( __children["SceneReader"] )
__children["SceneReader"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["distant_light"] = GafferArnold.ArnoldLight( "distant_light" )
parent.addChild( __children["distant_light"] )
__children["distant_light"].loadShader( "distant_light" )
__children["distant_light"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in3", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in4", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in5", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in6", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Camera"] = GafferScene.Camera( "Camera" )
parent.addChild( __children["Camera"] )
__children["Camera"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"] = GafferScene.Outputs( "Outputs" )
parent.addChild( __children["Outputs"] )
__children["Outputs"]["outputs"].addChild( Gaffer.ValuePlug( "output1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.StringPlug( "name", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.BoolPlug( "active", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.StringPlug( "fileName", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.StringPlug( "type", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.StringPlug( "data", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.CompoundDataPlug( "parameters", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.CompoundDataPlug.MemberPlug( "quantize", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["quantize"].addChild( Gaffer.StringPlug( "name", defaultValue = 'quantize', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["quantize"].addChild( Gaffer.IntVectorDataPlug( "value", defaultValue = IECore.IntVectorData( [ 0, 0, 0, 0 ] ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.CompoundDataPlug.MemberPlug( "displayHost", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["displayHost"].addChild( Gaffer.StringPlug( "name", defaultValue = 'displayHost', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["displayHost"].addChild( Gaffer.StringPlug( "value", defaultValue = 'localhost', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.CompoundDataPlug.MemberPlug( "remoteDisplayType", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["remoteDisplayType"].addChild( Gaffer.StringPlug( "name", defaultValue = 'remoteDisplayType', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["remoteDisplayType"].addChild( Gaffer.StringPlug( "value", defaultValue = 'GafferImage::GafferDisplayDriver', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.CompoundDataPlug.MemberPlug( "displayPort", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["displayPort"].addChild( Gaffer.StringPlug( "name", defaultValue = 'displayPort', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["displayPort"].addChild( Gaffer.StringPlug( "value", defaultValue = '${image:catalogue:port}', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.CompoundDataPlug.MemberPlug( "driverType", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["driverType"].addChild( Gaffer.StringPlug( "name", defaultValue = 'driverType', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"]["driverType"].addChild( Gaffer.StringPlug( "value", defaultValue = 'ClientDisplayDriver', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["InteractiveArnoldRender"] = GafferArnold.InteractiveArnoldRender( "InteractiveArnoldRender" )
parent.addChild( __children["InteractiveArnoldRender"] )
__children["InteractiveArnoldRender"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["StandardOptions"] = GafferScene.StandardOptions( "StandardOptions" )
parent.addChild( __children["StandardOptions"] )
__children["StandardOptions"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"] = GafferImage.Catalogue( "Catalogue" )
parent.addChild( __children["Catalogue"] )
__children["Catalogue"]["images"].addChild( GafferImage.Catalogue.Image( "Image", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"]["images"].addChild( GafferImage.Catalogue.Image( "Image1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"]["images"].addChild( GafferImage.Catalogue.Image( "Image2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"]["images"].addChild( GafferImage.Catalogue.Image( "Image3", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["standard_surface"] = GafferArnold.ArnoldShader( "standard_surface" )
parent.addChild( __children["standard_surface"] )
__children["standard_surface"].loadShader( "standard_surface" )
__children["standard_surface"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderAssignment"] = GafferScene.ShaderAssignment( "ShaderAssignment" )
parent.addChild( __children["ShaderAssignment"] )
__children["ShaderAssignment"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["skydome_light"] = GafferArnold.ArnoldLight( "skydome_light" )
parent.addChild( __children["skydome_light"] )
__children["skydome_light"].loadShader( "skydome_light" )
__children["skydome_light"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["physical_sky"] = GafferArnold.ArnoldShader( "physical_sky" )
parent.addChild( __children["physical_sky"] )
__children["physical_sky"].loadShader( "physical_sky" )
__children["physical_sky"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ArnoldOptions"] = GafferArnold.ArnoldOptions( "ArnoldOptions" )
parent.addChild( __children["ArnoldOptions"] )
__children["ArnoldOptions"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Plane"] = GafferScene.Plane( "Plane" )
parent.addChild( __children["Plane"] )
__children["Plane"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["light_blocker"] = GafferArnold.ArnoldLightFilter( "light_blocker" )
parent.addChild( __children["light_blocker"] )
__children["light_blocker"].loadShader( "light_blocker" )
__children["light_blocker"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["StandardAttributes"] = GafferScene.StandardAttributes( "StandardAttributes" )
parent.addChild( __children["StandardAttributes"] )
__children["StandardAttributes"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["PathFilter"] = GafferScene.PathFilter( "PathFilter" )
parent.addChild( __children["PathFilter"] )
__children["PathFilter"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["noise"] = GafferArnold.ArnoldShader( "noise" )
parent.addChild( __children["noise"] )
__children["noise"].loadShader( "noise" )
__children["noise"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SceneReader"]["fileName"].setValue( '/home/john/dev/build/gaffer/resources/gafferBot/caches/gafferBot.scc' )
__children["SceneReader"]["__uiPosition"].setValue( imath.V2f( 18.4374352, -6.75042582 ) )
__children["distant_light"]["enabled"].setValue( False )
__children["distant_light"]["sets"].setValue( 'keyLight' )
__children["distant_light"]["__uiPosition"].setValue( imath.V2f( 5.43743658, -6.7503953 ) )
__children["Group"]["in"]["in0"].setInput( __children["distant_light"]["out"] )
__children["Group"]["in"]["in1"].setInput( __children["SceneReader"]["out"] )
__children["Group"]["in"]["in2"].setInput( __children["Camera"]["out"] )
__children["Group"]["in"]["in3"].setInput( __children["skydome_light"]["out"] )
__children["Group"]["in"]["in4"].setInput( __children["Plane"]["out"] )
__children["Group"]["in"]["in5"].setInput( __children["light_blocker"]["out"] )
__children["Group"]["__uiPosition"].setValue( imath.V2f( 19.9374313, -14.9144878 ) )
__children["Camera"]["transform"]["translate"].setValue( imath.V3f( 16.3724556, 30.5729942, 42.350853 ) )
__children["Camera"]["transform"]["rotate"].setValue( imath.V3f( -21.7724056, 22.3453407, 5.12264205e-06 ) )
__children["Camera"]["__uiPosition"].setValue( imath.V2f( 31.4374352, -6.74875402 ) )
__children["Outputs"]["in"].setInput( __children["StandardOptions"]["out"] )
__children["Outputs"]["outputs"]["output1"]["name"].setValue( 'Interactive/Beauty' )
__children["Outputs"]["outputs"]["output1"]["fileName"].setValue( 'beauty' )
__children["Outputs"]["outputs"]["output1"]["type"].setValue( 'ieDisplay' )
__children["Outputs"]["outputs"]["output1"]["data"].setValue( 'rgba' )
__children["Outputs"]["__uiPosition"].setValue( imath.V2f( 19.9374313, -67.6603317 ) )
__children["InteractiveArnoldRender"]["in"].setInput( __children["Outputs"]["out"] )
__children["InteractiveArnoldRender"]["__uiPosition"].setValue( imath.V2f( 19.9374313, -75.8244095 ) )
__children["StandardOptions"]["in"].setInput( __children["ArnoldOptions"]["out"] )
__children["StandardOptions"]["options"]["renderCamera"]["value"].setValue( '/group/camera' )
__children["StandardOptions"]["options"]["renderCamera"]["enabled"].setValue( True )
__children["StandardOptions"]["__uiPosition"].setValue( imath.V2f( 19.9374313, -59.4962654 ) )
Gaffer.Metadata.registerValue( __children["Catalogue"]["images"]["Image"], 'image:index', 0 )
__children["Catalogue"]["images"]["Image"]["fileName"].setValue( '/home/john/gaffer/projects/default/catalogues/blockers/536cb8d8f89f509c24b25fc1cc101dae.exr' )
Gaffer.Metadata.registerValue( __children["Catalogue"]["images"]["Image1"], 'image:index', 1 )
__children["Catalogue"]["images"]["Image1"]["fileName"].setValue( '/home/john/gaffer/projects/default/catalogues/blockers/c057bf8191faaa76448f58bf0fb563c4.exr' )
Gaffer.Metadata.registerValue( __children["Catalogue"]["images"]["Image2"], 'image:index', 2 )
__children["Catalogue"]["images"]["Image2"]["fileName"].setValue( '/home/john/gaffer/projects/default/catalogues/blockers/7c01548619d08090ce4d87054a39accf.exr' )
Gaffer.Metadata.registerValue( __children["Catalogue"]["images"]["Image3"], 'image:index', 3 )
__children["Catalogue"]["images"]["Image3"]["fileName"].setValue( '/home/john/gaffer/projects/default/catalogues/blockers/62d792a53a6195a9dc90aaa48a5206d2.exr' )
__children["Catalogue"]["imageIndex"].setValue( 4 )
__children["Catalogue"]["directory"].setValue( '${project:rootDirectory}/catalogues/${script:name}' )
__children["Catalogue"]["__mapping"].setValue( IECore.CompoundData({'Image':IECore.IntData( 0 ),'Image1':IECore.IntData( 1 ),'Image2':IECore.IntData( 2 ),'Image3':IECore.IntData( 3 ),'Image4':IECore.IntData( 4 )}) )
__children["Catalogue"]["__uiPosition"].setValue( imath.V2f( 41.0015259, -75.1076279 ) )
__children["standard_surface"]["__uiPosition"].setValue( imath.V2f( -8.97600555, -47.3448639 ) )
__children["ShaderAssignment"]["in"].setInput( __children["StandardAttributes"]["out"] )
__children["ShaderAssignment"]["shader"].setInput( __children["standard_surface"]["out"] )
__children["ShaderAssignment"]["__uiPosition"].setValue( imath.V2f( 19.9374313, -47.9472809 ) )
__children["skydome_light"]["transform"]["rotate"].setValue( imath.V3f( 0, -48.2799988, 0 ) )
__children["skydome_light"]["parameters"]["color"].setInput( __children["physical_sky"]["out"] )
__children["skydome_light"]["parameters"]["color"]["r"].setInput( __children["physical_sky"]["out"]["r"] )
__children["skydome_light"]["parameters"]["color"]["g"].setInput( __children["physical_sky"]["out"]["g"] )
__children["skydome_light"]["parameters"]["color"]["b"].setInput( __children["physical_sky"]["out"]["b"] )
__children["skydome_light"]["parameters"]["camera"].setValue( 0.0 )
__children["skydome_light"]["__uiPosition"].setValue( imath.V2f( 46.5734253, -6.75042534 ) )
__children["physical_sky"]["__uiPosition"].setValue( imath.V2f( 32.0785522, 2.96439648 ) )
__children["ArnoldOptions"]["in"].setInput( __children["ShaderAssignment"]["out"] )
__children["ArnoldOptions"]["options"]["aaSamples"]["enabled"].setValue( True )
__children["ArnoldOptions"]["__uiPosition"].setValue( imath.V2f( 19.9374313, -53.8322029 ) )
__children["Plane"]["transform"]["rotate"].setValue( imath.V3f( 270, 0, 0 ) )
__children["Plane"]["transform"]["scale"]["x"].setValue( 70167.46875 )
__children["Plane"]["transform"]["scale"]["y"].setInput( __children["Plane"]["transform"]["scale"]["x"] )
__children["Plane"]["transform"]["scale"]["z"].setInput( __children["Plane"]["transform"]["scale"]["x"] )
__children["Plane"]["__uiPosition"].setValue( imath.V2f( 61.2554169, -6.75042534 ) )
__children["light_blocker"]["transform"]["translate"].setValue( imath.V3f( -3.88295555, 5.37757587, 0 ) )
__children["light_blocker"]["transform"]["scale"].setValue( imath.V3f( 28.9176025, 9.44146633, 28.9176025 ) )
__children["light_blocker"]["parameters"]["geometry_type"].setValue( 'sphere' )
__children["light_blocker"]["parameters"]["density"].setValue( 1.0 )
__children["light_blocker"]["parameters"]["width_edge"].setValue( 1.100000023841858 )
__children["light_blocker"]["parameters"]["height_edge"].setValue( 0.20000000298023224 )
__children["light_blocker"]["parameters"]["axis"].setValue( 'y' )
__children["light_blocker"]["parameters"]["shader"].setInput( __children["noise"]["out"] )
__children["light_blocker"]["parameters"]["shader"]["r"].setInput( __children["noise"]["out"]["r"] )
__children["light_blocker"]["parameters"]["shader"]["g"].setInput( __children["noise"]["out"]["g"] )
__children["light_blocker"]["parameters"]["shader"]["b"].setInput( __children["noise"]["out"]["b"] )
__children["light_blocker"]["__uiPosition"].setValue( imath.V2f( 102.970245, -6.75042534 ) )
__children["StandardAttributes"]["in"].setInput( __children["Group"]["out"] )
__children["StandardAttributes"]["filter"].setInput( __children["PathFilter"]["out"] )
__children["StandardAttributes"]["attributes"]["linkedLights"]["enabled"].setValue( True )
__children["StandardAttributes"]["attributes"]["filteredLights"]["value"].setValue( 'defaultLights' )
__children["StandardAttributes"]["attributes"]["filteredLights"]["enabled"].setValue( True )
__children["StandardAttributes"]["__uiPosition"].setValue( imath.V2f( 19.9374313, -29.6995449 ) )
__children["PathFilter"]["paths"].setValue( IECore.StringVectorData( [ '/group/lightFilter' ] ) )
__children["PathFilter"]["__uiPosition"].setValue( imath.V2f( 34.3427048, -23.6166534 ) )
__children["noise"]["__uiPosition"].setValue( imath.V2f( 82.6487427, 2.23191619 ) )


del __children


```

mattigruener replied:
Ah, I haven't explained enough what I'm doing either :P I see where the differences we're seeing are coming from now. Try using an aiImage and see what happens then. But even when using a noise, the ramp _parameter_ still has an effect, no?

johnhaddon replied:
Ah, yes, so aiImage doesn't behave except when using the box shape. I _think_ this is because it's relying on UVs being specified, and only the box mode seems to be specifying them. In other geometry modes, I can use the `aiImage.offset` parameter to shift around the UVs, entering `0.5, 0.5` to sample from the middle of the image, implying that those modes are simply setting the UVs to a constant `0,0`.

When using a noise, the `ramp` parameter is having an effect, yes. And the shape of that ramp is even dictated by the `geometry_type` parameter. But the shape itself isn't used to mask the result of the shader, and the `width_edge` and `width_height` parameters don't appear to be doing anything. Except when in box mode of course, where they adjust the UVs that the shader uses, presumably so that `0-1` fills the full size of the box, all the way to the soft edge.

I have no idea how to explain this to a user. Perhaps this?

- Disable `width_edge` and `width_height` when a shader is connected and the mode isn't box.
- Add a tooltip saying that UVs are only available to the shader in Box mode, so you'll need to shade based on P.

johnhaddon commented on python/GafferArnoldUI/ArnoldLightFilterUI.py:
```diff
@@ -91,6 +100,13 @@ def __activateRoundness( parent ) :
 
+			"description",
+			"""
+			Shader to be used for the light_blocker filter. Only works if geometry type
```
I'm seeing the opposite - it always works, but it causes the geometry type to be ignored.

mattigruener replied:
Really? If you set the type to "sphere" and plug in a shader, you still see some filtering happening? From the Arnold docs: "It only works when 'Geometry Type' is set to Box."

johnhaddon replied:
Yeah, it's like the shader just provides the final filter value (example pasted in previous comment).

johnhaddon commented:
I'm baffled by the Travis failure on Mac BTW - currently trying to reproduce on my local Mac.

johnhaddon commented:
>I'm baffled by the Travis failure on Mac BTW - currently trying to reproduce on my local Mac.

Turned out to be a trivial linking issue - you'll just need this diff :

```
--- a/SConstruct
+++ b/SConstruct
@@ -782,7 +782,7 @@ libraries = {
 
        "GafferArnoldUI" : {
                "envAppends" : {
-                       "LIBS" : [ "IECoreScene$CORTEX_LIB_SUFFIX", "IECoreGL$CORTEX_LIB_SUFFIX", "GafferOSL", "GafferSceneUI" ],
+                       "LIBS" : [ "IECoreScene$CORTEX_LIB_SUFFIX", "IECoreGL$CORTEX_LIB_SUFFIX", "Gaffer", "GafferOSL", "GafferSceneUI" ],
                        },
                "pythonEnvAppends" : {
                        "LIBS" : [ "GafferArnoldUI", "GafferSceneUI" ],
```

mattigruener commented:
Hey John,

thanks for your comments :)

> perhaps the "filteredLights" plug should just go under the "Light Linking" section anyway. They seem pretty strongly related?

Yeah, I wasn't sure about where to put it. I've come to think of the two mechanisms as somewhat separate, but maybe that's wrong. I'm just thinking that it'd add some confusion if we mix the two things too much. I do agree that they're strongly related, but that's exactly why I can see people using the light linking stuff to hook up their filters and we'd then tell them that that's not how it works and that _light linking_ is something else, but that they should use this other plug in the _Light Linking_ section.  
Sections with only one plug in them are pretty ugly too, though. We can try to clarify things in the tooltips, but I'm not sure people actually read them?

> Have you done any stress testing of this with large numbers of lights/blockers/links, like we did for the light linking work?

I haven't. I'll look into it!

I'll comment on the rest inline. Thanks! :)

Matti.

johnhaddon commented:
> Yeah, I wasn't sure about where to put it. I've come to think of the two mechanisms as somewhat separate, but maybe that's wrong. I'm just thinking that it'd add some confusion if we mix the two things too much. I do agree that they're strongly related, but that's exactly why I can see people using the light linking stuff to hook up their filters and we'd then tell them that that's not how it works and that light linking is something else, but that they should use this other plug in the Light Linking section.
Sections with only one plug in them are pretty ugly too, though. We can try to clarify things in the tooltips, but I'm not sure people actually read them?

Let's gather opinions from the Murrays of the world to help inform the decision. My opinion is tentatively that a single Light Linking section is preferable, and that it would go well with the proposed "Light Linking" tab on the filter itself. Both "linkedLights" and "filteredLights" establish a connection between objects and  lights, and both accept set expressions in the same format. I agree there is some potential for confusion between the two, but I almost think the distinction is better made by placing them side by side, so you always see them at the same time. That way you can't accidentally use the wrong one without even seeing the other. And collapsibles with just one thing inside are a bit ugly.

mattigruener commented:
I think I addressed what I can at the moment. The Sconstruct fix is in a7cf8f9.

Let's discuss what to do about the UI regarding the `shader` input connections. That could get quite messy... :/ I sent Murray a message about the other outstanding UI issue.

johnhaddon commented:
Thanks Matti! Some interested parties at Cinesite would quite like to test this before the final release of 0.54, so once you've verified that performance is sound, I propose we merge and make a preview build for folks to try out. Then we can address any further comments before committing to a release...

mattigruener commented:
I added the threading fix in a separate commit.

mattigruener commented:
Hey John,

I've added a few commits that add quite significant optimizations for both memory usage and speed. The respective numbers are in the shared spreadsheet we've been working with.

I've kept everything in separate commits for now because I think it's easier to think through the individual steps that way. Maybe this also calls for another video call? I'm guessing that there'll be a few things here that'll need changing, but hopefully the overall approach is alright... Previously we've spent _a lot_ of time spinning and now things seem nice and parallel.

Also added the commit that fixes the crashes CS was seeing (daa4f14) and one that moves the `filteredLights` plug as discussed (8ced504). There's a new test that deals with larger amounts of lights and filters now, too (826f01b). Eventually that one can be made into an automatic performance test.

Cheerio!

Matti.

mattigruener commented:
Cleaned up the commit history, but left the optimization related work untouched for easier reviewing. Updated the commit hashes in my previous comment, too.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -1933,6 +1972,197 @@ IE_CORE_FORWARDDECLARE( ArnoldObject )
+	};
+
+	using FilterGroupPtr = std::shared_ptr<FilterGroup>;
+	using FilterGroups = tbb::concurrent_hash_map<const IECore::StringVectorData*, FilterGroupPtr>;
```
Is `StringVectorData *` the right key here? I think we're OK in terms of object identity and lifetimes because the ArnoldLightFilter class keeps the data alive via ArnoldAttributes? But for performance we're relying a lot on the Renderer client using the same exact object for each distinct set of lights aren't we? I guess Gaffer's caching of EvaluateLightLinks is definitely on our side there, but I wonder if we should use `filteredLightsData->hash()` as the key instead?

mattigruener replied:
Those were my thoughts, as well, and I concluded that things would work out alright, but I've also felt that using the raw pointer here is dodgy and will look dangerous to everyone else reading the code. I'm happy to use the hash instead. Probably worth running another performance test with the hashing in place, though?

johnhaddon replied:
The hash should be cheap enough I think - it's stored internal to the StringVectorData so won't be computed repeatedly unless we have different objects. Always good to measure again though...

mattigruener replied:
I had missed that I'm actually using the pointer in `LightFilterConnections::update()` to loop through all lights that are affected by the FilterGroup. I'll keep things the way they are for now as working around this problem would probably obfuscate the code more.

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -2087,13 +2350,238 @@ class ArnoldLight : public ArnoldObject
+		return;
+	}
+
+	if( m_ownConnections )
```
That sounds good to me, thanks for the suggestion.

mattigruener commented:
Addressed 1/2 comments. Resolved merge conflict.

Do you want me to squash in the optimizations? I thought it would be worth having a record of simpler times and the step-by-step bits towards a faster implementation.

johnhaddon commented:
:cry:

```
src/GafferScene/StandardAttributes.cpp: In constructor 'GafferScene::StandardAttributes::StandardAttributes(const string&)':
src/GafferScene/StandardAttributes.cpp:70:14: error: 'class Gaffer::CompoundDataPlug' has no member named 'addOptionalMember'
  attributes->addOptionalMember( "filteredLights", new IECore::StringData( "" ), "filteredLights", Gaffer::Plug::Default, false );
```

Should be a simple fix though - `addOptionalMember()` got removed in the `NameValuePlug` PR, so you just need to match the surrounding calls which have been updated as needed.

mattigruener commented:
Pulled in public changes, rebased, fixed the issue, squashed in the changes, updated the PR.

I've also cleaned up the history a tiny bit more.

johnhaddon commented:
Restarted the Linux tests - for some reason the Travis VMs didn't get past the package installation step. Definitely nothing to do with the PR.

johnhaddon commented:
Tests have passed - merging. PR #3020, rest in peace.

mattigruener commented:
Thanks John :) May this be helpful to people. I think it will be, I've been asked about blockers again yesterday. Some people are keen on getting access to them.
